src/event/ngx_event_openssl.c

```
@@ -1315,6 +1315,8 @@ ngx_ssl_passwords_cleanup(void *data)
 ngx_int_t
 ngx_ssl_dhparam(ngx_conf_t *cf, ngx_ssl_t *ssl, ngx_str_t *file)
 {
+#ifndef OPENSSL_NO_DH
```

Ah, yes. That would be more correct. Despite having `SSL_CTX_set0_tmp_dh_pkey` available, we don't have an `EVP_PKEY_ASN1_METHOD` instance registered for parsing `DH PARAMETERS` and cannot succeed anyways.

I wonder if we should treat this as an error. The original patch resulted in this nice message:
```
nginx: [emerg] PEM_read_bio_Parameters("/tmp/nginx-test-CR8zJrHShe/dhparam.pem") failed (SSL: error:1E08010C:DECODER routines::unsupported:No supported data to decode. Input type: PEM)
```

---

Yes, that's the question.

I followed what we do with ecdhe curve (the next function) in the absence of ecdh support (no-ecdh).

This has a good explanation:
given that both no-dh and no-ecdh are opt in, it is your responsibility to provide a sensible configuration.

Also, and this applies to curves in some sense as well, this directive provides DHE parameters to enable ephemeral DH ciphers, which are otherwise disabled. DHE ciphers were generally substituted with elliptic curves maybe around 10 years ago (but see RFC 7919).
So I somewhat perceive this as a legacy.

Given the above I think it's fine to fail silently.